### PR TITLE
Ignore objects with names that end with / when listing rule groups

### PR DESCRIPTION
**Pull request by pstibrany**
**What this PR does**: This PR changes bucket-client based rule-store implementation to ignore objects that end with `/`. Such objects may exist and would be returned, but would cause failure to load the rule later, which can lead to ruler not loading any rules.

**Checklist**
- [x] Tests updated
- [ ] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -21,6 +21,7 @@
 * [BUGFIX] Distributor: reverted changes done to rate limiting in #3825. #3948
 * [BUGFIX] Ingester: Fix race condition when opening and closing tsdb concurrently. #3959
 * [BUGFIX] Querier: streamline tracing spans. #3924
+* [BUGFIX] Experimental Ruler Storage: ignore objects with empty namespace or group in the name. #3999
```
Comment (pracucci) [inline CHANGELOG.md:24]: No more experimental ;)

```suggestion
* [BUGFIX] Ruler Storage: ignore objects with empty namespace or group in the name. #3999
```

**Review by pracucci — APPROVED** [review]
Thanks for the fix! LGMT

**Review by pracucci — COMMENTED** [review]
I approved, but I'm doing a step back now. `/` is a valid base64 character, so we may have it within the key name as well, which would break the logic anyway. Also having the key with `/` at the end may be legit 🤔 

Am I missing anything?

**Comment by pstibrany** [comment]
> I approved, but I'm doing a step back now. `/` is a valid base64 character, so we may have it within the key name as well, which would break the logic anyway. Also having the key with `/` at the end may be legit 🤔
> 
> Am I missing anything?

We use base64 with "URL" alphabet (`ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_`), so `/` is not a valid character.

Key with `/` is legal in general, but our rule group objects must have non-empty name, which implies that they cannot have name ending with `/`.

**Comment by pracucci** [comment]
> We use base64 with "URL" alphabet (ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_), so / is not a valid character.

Interesting. So, how can we end up with an object ending with `/`?

**Comment by pstibrany** [comment]
> Interesting. So, how can we end up with an object ending with `/`?

Objects that we've found are from Aug 2020 and Dec 2020. I suspect that we didn't check for group or namespace names in the past (do we now?). Content of those files is suspicious too -- they don't look like rule group files at all.

**Comment by pracucci** [comment]
> Objects that we've found are from Aug 2020 and Dec 2020. I suspect that we didn't check for group or namespace names in the past (do we now?). Content of those files is suspicious too -- they don't look like rule group files at all.

My though is: if such objects are not supposed to be there, should we filter them out (this PR)? I would bet we can break several parts of cortex if the storage contains unexpected objects.

File: pkg/ruler/rulestore/bucketclient/bucket_client.go [hunk 2]
```diff
@@ -102,9 +102,16 @@ func (b *BucketRuleStore) ListAllRuleGroups(ctx context.Context) (map[string]rul
...
 	// List rule groups for all tenants.
 	err := b.bucket.Iter(ctx, "", func(key string) error {
+		// Ignore objects that look like directories, ie. they have empty namespace or group name.
+		// Such objects can exist in the object store, but would cause trouble when loading them, because
+		// we check for user, namespace and group components not being empty.
+		if strings.HasSuffix(key, objstore.DirDelim) {
+			return nil
+		}
+
 		userID, namespace, group, err := parseRuleGroupObjectKeyWithUser(key)
```
Comment (pracucci) [inline pkg/ruler/rulestore/bucketclient/bucket_client.go:112]: What if we actually change `parseRuleGroupObjectKeyWithUser()` and return an error if the group is empty?

Comment (pstibrany) [reply pkg/ruler/rulestore/bucketclient/bucket_client.go:112]: We should really check for user, namespace or group being non-empty.

Comment (pracucci) [reply pkg/ruler/rulestore/bucketclient/bucket_client.go:108]: I thought it wasn't possible to have empty user and namespace due to how the parsing work but, if it's possible, then definitely yet.

**Comment by pstibrany** [comment]
Note that validation for empty group names was added in https://github.com/cortexproject/cortex/pull/3210.

**Review by pracucci — APPROVED** [review]
Looks very good to me, thanks!

**Pull request merged**